kvdb/postgres/db.go

```diff
@@ -0,0 +1,243 @@
+		}
+	}()
+
+	err = f()
+
+	return
+}
+
+// View opens a database read transaction and executes the function f with the
+// transaction passed as a parameter. After f exits, the transaction is rolled
+// back. If f errors, its error is returned, not a rollback error (if any
+// occur). The passed reset function is called before the start of the
+// transaction and can be used to reset intermediate state. As callers may
+// expect retries of the f closure (depending on the database backend used), the
+// reset function will be called before each retry respectively.
+func (db *db) View(f func(tx walletdb.ReadTx) error, reset func()) error {
+	return db.executeTransaction(
+		func(tx walletdb.ReadWriteTx) error {
+			return f(tx.(walletdb.ReadTx))
+		},
+		reset, true,
+	)
+}
+
+// Update opens a database read/write transaction and executes the function f
+// with the transaction passed as a parameter. After f exits, if f did not
+// error, the transaction is committed. Otherwise, if f did error, the
+// transaction is rolled back. If the rollback fails, the original error
+// returned by f is still returned. If the commit fails, the commit error is
+// returned. As callers may expect retries of the f closure, the reset function
+// will be called before each retry respectively.
+func (db *db) Update(f func(tx walletdb.ReadWriteTx) error, reset func()) (err error) {
+	return db.executeTransaction(f, reset, false)
+}
+
+// executeTransaction creates a new read-only or read-write transaction and
+// executes the given function within it.
+func (db *db) executeTransaction(f func(tx walletdb.ReadWriteTx) error,
+	reset func(), readOnly bool) error {
+
+	reset()
+
+	tx, err := newReadWriteTx(db, readOnly)
+	if err != nil {
+		return err
+	}
+
+	err = catchPanic(func() error { return f(tx) })
+	if err != nil {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			log.Errorf("Error rolling back tx: %v", rollbackErr)
+		}
+
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// PrintStats returns all collected stats pretty printed into a string.
+func (db *db) PrintStats() string {
+	return "stats not supported by Postgres driver"
+}
+
+// BeginReadWriteTx opens a database read+write transaction.
+func (db *db) BeginReadWriteTx() (walletdb.ReadWriteTx, error) {
+	return newReadWriteTx(db, false)
+}
+
+// BeginReadTx opens a database read transaction.
+func (db *db) BeginReadTx() (walletdb.ReadTx, error) {
+	return newReadWriteTx(db, true)
+}
+
+// Copy writes a copy of the database to the provided writer. This call will
+// start a read-only transaction to perform all operations.
+// This function is part of the walletdb.Db interface implementation.
+func (db *db) Copy(w io.Writer) error {
+	return errors.New("not implemented")
+}
+
+// Close cleanly shuts down the database and syncs all data.
+// This function is part of the walletdb.Db interface implementation.
+func (db *db) Close() error {
+	db.db.Close()
```

Comment:
Return the error from `db.db.Close()` here?

Reply:
Fixed